src/django_components/component.py

```diff
@@ -3645,26 +3669,70 @@ def _render_impl(
         # Cleanup
         context.render_context.pop()  # type: ignore[union-attr]
 
+        trace_component_msg(
+            "COMP_PREP_END",
+            component_name=component_name,
+            component_id=render_id,
+            slot_name=None,
+            component_path=component_path,
+        )
+
         ######################################
         # 5. Render component
         #
         # NOTE: To support infinite recursion, we don't directly call `Template.render()`.
-        #       Instead, we defer rendering of the component - we prepare a callback that will
-        #       be called when the rendering process reaches this component.
+        #       Instead, we defer rendering of the component - we prepare a generator function
+        #       that will be called when the rendering process reaches this component.
         ######################################
 
+        trace_component_msg(
+            "COMP_RENDER_START",
+            component_name=component.name,
+            component_id=component.id,
+            slot_name=None,
+            component_path=component_path,
+        )
+
+        component.on_render_before(context_snapshot, template)
+
+        # Emit signal that the template is about to be rendered
+        if template is not None:
+            template_rendered.send(sender=template, template=template, context=context_snapshot)
+
         # Instead of rendering component at the time we come across the `{% component %}` tag
         # in the template, we defer rendering in order to scalably handle deeply nested components.
         #
-        # See `_gen_component_renderer()` for more details.
-        deferred_render = component._gen_component_renderer(
+        # See `_make_renderer_generator()` for more details.
+        renderer_generator = component._make_renderer_generator(
             template=template,
             context=context_snapshot,
             component_path=component_path,
-            css_input_hash=css_input_hash,
-            js_input_hash=js_input_hash,
         )
 
+        # This callback is called with the value that was yielded from `Component.on_render()`.
+        # It may be called multiple times for the same component, e.g. if `Component.on_render()`
+        # contains multiple `yield` keywords.
+        def on_component_intermediate(html_content: Optional[str]) -> Optional[str]:
+            # HTML attributes passed from parent to current component.
+            # NOTE: Is `None` for the root component.
+            curr_comp_attrs = component_tree_context.component_attrs.get(render_id, None)
+
+            if html_content:
```

Comment:
Extensions like CSS scoping will be able to modify the "intermediate" component result here. In fact, the this `if html_content` branch could be actually extracted out as an internal extension too.